ext/bcmath/bcmath.c

```diff
@@ -777,4 +791,943 @@ PHP_FUNCTION(bcscale)
 /* }}} */
 
 
+
+static zend_class_entry *bcmath_number_ce;
+static zend_object_handlers bcmath_number_obj_handlers;
+
+static zend_result bcmath_number_do_operation(uint8_t opcode, zval *ret_val, zval *op1, zval *op2);
+static int bcmath_number_compare(zval *op1, zval *op2);
+
+#if SIZEOF_SIZE_T >= 8
+#  define CHECK_RET_SCALE_OVERFLOW(scale, origin_scale) (scale > INT_MAX)
+#else
+#  define CHECK_RET_SCALE_OVERFLOW(scale, origin_scale) (scale > INT_MAX || scale < origin_scale)
+#endif
+
+static zend_always_inline bcmath_number_obj_t *get_bcmath_number_from_obj(const zend_object *obj)
+{
+	return (bcmath_number_obj_t*)((char*)(obj) - XtOffsetOf(bcmath_number_obj_t, std));
+}
+
+static zend_always_inline bcmath_number_obj_t *get_bcmath_number_from_zval(const zval *zv)
+{
+	return get_bcmath_number_from_obj(Z_OBJ_P(zv));
+}
+
+static zend_always_inline zend_string *bcmath_number_value_to_str(bcmath_number_obj_t *intern)
+{
+	if (intern->value == NULL) {
+		intern->value = bc_num2str_ex(intern->num, intern->scale);
+	}
+	return intern->value;
+}
+
+static zend_object *bcmath_number_create(zend_class_entry *ce)
+{
+	bcmath_number_obj_t *intern = zend_object_alloc(sizeof(bcmath_number_obj_t), ce);
+
+	zend_object_std_init(&intern->std, ce);
+	object_properties_init(&intern->std, ce);
+
+	intern->scale = 1;
+
+	return &intern->std;
+}
+
+static void bcmath_number_free(zend_object *obj)
+{
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	if (intern->num) {
+		bc_free_num(&intern->num);
+		intern->num = NULL;
+	}
+	if (intern->value) {
+		zend_string_release(intern->value);
+		intern->value = NULL;
+	}
+	zend_object_std_dtor(&intern->std);
+}
+
+static zend_object *bcmath_number_clone(zend_object *obj)
+{
+	bcmath_number_obj_t *original = get_bcmath_number_from_obj(obj);
+	bcmath_number_obj_t *clone = get_bcmath_number_from_obj(bcmath_number_create(bcmath_number_ce));
+
+	clone->num = bc_copy_num(original->num);
+	if (original->value) {
+		clone->value = zend_string_copy(original->value);
+	}
+	clone->scale = original->scale;
+
+	return &clone->std;
+}
+
+static HashTable *bcmath_number_get_properties_for(zend_object *obj, zend_prop_purpose purpose)
+{
+	zval zv;
+	bcmath_number_obj_t *intern = get_bcmath_number_from_obj(obj);
+	HashTable *props = zend_array_dup(zend_std_get_properties(obj));
+
+	ZVAL_STR_COPY(&zv, bcmath_number_value_to_str(intern));
+	zend_hash_str_update(props, ZEND_STRL("value"), &zv);
```

Comment:
Just found out that "value" is a "known string", so it's cheaper to do it like this:

```suggestion
	zend_hash_update(props, ZSTR_KNOWN(ZEND_STR_VALUE), &zv);
```